microsetta_admin/server.py

```diff
@@ -774,6 +776,248 @@ def scan():
+                if val is None:
+                    tmp = legends['None']
+                    curr_cols.append([code, tmp, proj, status, src])
+                else:
+                    curr_cols.append([code, legends[val], proj, status, src])
+
+    output_dict[curr_row] = curr_cols
+    return output_dict
+
+
+def _visualize_scans():
+    filename = request.form['scanned_file']
+    criteria = request.form['sort_criteria']
+    session['recent_criteria'] = criteria
+    legends = _get_legends(criteria)
+
+    return render_template('bulk_scan.html',
+                           **build_login_variables(),
+                           stage="Visualization",
+                           filename=filename,
+                           data=criteria,
+                           legends=legends,
+                           table=_compose_table(legends, criteria),
+                           status_options=STATUS_OPTIONS,
+                           message=""
+                           )
+
+
+def _post_bulk_scan_add():
+    send_email = request.form.get('send_email', False)
+    session[SEND_EMAIL_CHECKBOX_DEFAULT_NAME] = send_email
+
+    sample_barcode = request.form['sample_barcode']
+    technician_notes = request.form['technician_notes']
+    sample_status = request.form['sample_status']
+
+    # Do the actual update
+    status, response = APIRequest.post(
+        '/api/admin/scan/%s' % sample_barcode,
+        json={
+            "sample_status": sample_status,
+            "technician_notes": technician_notes
+        }
+    )
+
+    message = ""
+    if status == 201:
+        message = "Success: Sample scanned successfully!"
+    else:
+        message = "Error: Unable to scan sample!"
+
+    filename = session.get('scan_file', "")
+    criteria = session.get('recent_criteria', "Project")
+    legends = _get_legends(criteria)
+    return render_template('bulk_scan.html',
+                           **build_login_variables(),
+                           stage="Visualization",
+                           filename=filename,
+                           data=criteria,
+                           legends=legends,
+                           table=_compose_table(legends, criteria),
+                           status_options=STATUS_OPTIONS,
+                           message=message
+                           )
+
+
+def _post_bulk_scan():
+    obj_file = request.files['file_picker']
+    sort_criteria = request.form['sort_criteria']
+
+    if obj_file is not None:
+        filename = obj_file.filename
+        ds = pd.read_csv(obj_file, dtype={'TubeCode': str, 'RackID': str})
+        file_data = ds.values.tolist()
+
+        scanned_samples = []
+        for rec in file_data:
+            status, response = APIRequest.get(
+                '/api/admin/rack/sample/%s' % rec[6],)
+
+            if status == 404:
+                obj = {}
+                obj["rack_id"] = rec[6]
+                if math.isnan(rec[3]):
```

Comment:
If the positional information is missing from the file, it should throw an error, not silently allow a null value.